dag_in_context/src/utility/context-prop.egg

```diff
@@ -1,49 +1,69 @@
 ;; Besides the in_context rules that add fresh contexts to everything,
-;; we normalize terms to keep contexts in the leaf nodes.
-;; This file contains the rules that move contexts to leaf nodes
-
+;; we normalize terms by pushing loop and func contexts to leaf nodes.
+;; Pushing InIf contexts down to leaves causes exponential blowup, so we avoid this.
+ 
 (ruleset context-prop)
 
+;; A leaf context is either a InLoop or a NoContext.
+;; We maintain the invariant that there is only one such context per region, and it can be pushed down to
```

Comment:
Should we also add rules to check this invariant? Would the commented-out rule in `context_of.egg` check this invariant?